sycl/test/basic_tests/buffer/subbuffer_data.cpp

```diff
@@ -0,0 +1,211 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple  %s -o %t.out
+// RUN: env SYCL_PI_TRACE=2 SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: env SYCL_PI_TRACE=2 %CPU_RUN_PLACEHOLDER %t.out
+// RUN: env SYCL_PI_TRACE=2 %GPU_RUN_PLACEHOLDER %t.out
+// RUN: env SYCL_PI_TRACE=2 %ACC_RUN_PLACEHOLDER %t.out
+//
+//==---------- subbuffer_data.cpp --- sub-buffer test for data copy back and destruction ---------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+using namespace cl::sycl;
+
+constexpr long total = 1'024;
+constexpr long subSz = total / 2;
+
+// setup_arr makes data go 1, 2, ... n. (like iota)
+void setup_arr(int *arr) {
+  for (long i = 0; i < total; i++) {
+    arr[i] = i;
+  }
+}
+
+// clear_arr sets all data to 0.
+void clear_arr(int *arr, int offset = 0) {
+  for (long i = offset; i < total; i++) {
+    arr[i] = 0;
+  }
+}
+
+void report_arr(int *arr, std::string msg) {
+  std::cout << ":: " << msg << " ::" << std::endl;
+  for (long i = 0; i < total; i++) {
+    std::cout << i << ": " << arr[i] << std::endl;
+  }
+}
+
+void ensureNoUnecessaryCopyBack(queue &q) {
+
+  std::cout << "start ensureNoUnecessaryCopyBack" << std::endl;
+
+  //allocate  memory
+  int *baseData = (int *)(malloc(total * sizeof(int)));
+  int *otherData = (int *)(malloc(total * sizeof(int)));
+
+  // ------- basic copy back ------------
+  //  two buffers are setup, one is read from, one is written to.
+  //  When ~dtor, only the written-to buffer should have been copied back.
+  //  We test this by clearing the read buffer behind its back and then check its value after ~dtor,
+  //  to make sure that it was not copied back
+  //  SYCL guarantees that the write buffer will have copied back no later than its ~dtor, but, in fact,
+  //  it might copy back earlier than that (and almost certainly will for CPU bound devices).
+  //  But we don't care about the write buffer. We only care that the read buffer was NOT copied-back.
+  { // closure
+    //setup and clear memory
+    setup_arr(baseData);  // [0, 1, 2,  ..., total]
+    clear_arr(otherData); // [0, 0, 0, ..., 0]
+    //buffers
+    buffer<int, 1> readFrom(baseData, range<1>(total));
+    buffer<int, 1> writeTo(otherData, range<1>(total));
+
+    q.submit([&](handler &cgh) {
+      auto readAcc = readFrom.get_access<access::mode::read>(cgh);
+      auto writeAcc = writeTo.get_access<access::mode::write>(cgh);
+      cgh.parallel_for<class noCopyBackRead>(range<1>(total), [=](id<1> i) {
+        writeAcc[i] = readAcc[i] + 1;
+      });
+    });
+    q.wait();
+
+    // change readBuffer behind its back.
+    clear_arr(baseData);
```

Comment:
In the[SYCL 2020 provisional specs](https://www.khronos.org/registry/SYCL/specs/sycl-2020-provisional.pdf), it says on page 123, second paragraph that "2. A buffer can be constructed with associated host memory and a default buffer allocator. The buffer will use
this host memory for its full lifetime, but the contents of this host memory are unspecified for the lifetime
of the buffer. If the host memory is modified by the host, or mapped to another buffer or image during the
lifetime of this buffer, then the results are undefined. The initial contents of the buffer will be the contents
of the host memory at the time of construction."

Does this mean that this test has undefined behaviour by SYCL specs because it doesn't use an accessor to access and modify host data that is still in use by a buffer?

I am also wondering whether it is still an undefined behavior to read the host data when its buffer is only accessed via read-accessors throughout its lifetime. The specs say that it is undefined behavior when we try to modify it but it doesn't state whether my use case is allowed or not.

Reply:
It's UB to access host data passed to the buffer c'tor until the buffer is destroyed.